ocis-pkg/runner/grouprunner.go

```diff
@@ -0,0 +1,130 @@
+// more or less independent from eachother, but at the same time it doesn't
+// make sense to have any of them stopped while the rest are running.
+// Basically, either all of them run, or none of them.
+// For example, you can have a GRPC and HTTP servers running, each of them
+// providing a piece of functionality, however, if any of them fails, the
+// feature provided by them would be incomplete or broken.
+//
+// Note that, as services, the task aren't expected to stop by default.
+// This means that, if a task finishes naturally, the rest of the task will
+// asked to stop as well.
+type GroupRunner struct {
+	runners []*Runner
+}
+
+// NewGroupRunner will create a GroupRunner
+func NewGroupRunner() *GroupRunner {
+	return &GroupRunner{
+		runners: []*Runner{},
+	}
+}
+
+// Add will add a runner to the group.
+//
+// It's mandatory that each runner in the group has an unique id, otherwise
+// there will be issues
+func (gr *GroupRunner) Add(r *Runner) {
+	gr.runners = append(gr.runners, r)
+}
+
+// Run will execute all the tasks in the group at the same time.
+//
+// Similarly to the "regular" runner's `Run` method, the execution thread
+// will be blocked here until all tasks are completed, and their results
+// will be available (each result will have the runner's id so it's easy to
+// find which one failed). Note that there is no guarantee about the result's
+// order, so the first result in the slice might or might not be the first
+// result to be obtained.
+//
+// When the context is marked as done, the groupRunner will call all the
+// stoppers for each runner to notify each task to stop. Note that the tasks
+// might still take a while to complete.
+//
+// If a task finishes naturally (with the context still "alive"), it will also
+// cause the groupRunner to call the stoppers of the rest of the tasks. So if
+// a task finishes, the rest will also finish.
+// Note that it is NOT expected for the finished task's stopper to be called
+// in this case.
+func (gr *GroupRunner) Run(ctx context.Context) []*Result {
+	results := make(map[string]*Result)
+
+	ch := make(chan *Result, len(gr.runners)) // no need to block writing results
+	for _, runner := range gr.runners {
+		runner.RunNoContext(ch)
+	}
+
+	// wait for a result or for the context to be done
+	select {
+	case result := <-ch:
+		results[result.RunnerID] = result
+	case <-ctx.Done():
+		// Do nothing
+	}
+
+	// interrupt the rest of the runners
+	for _, runner := range gr.runners {
+		if _, ok := results[runner.ID]; !ok {
+			// there might still be race conditions because the result might not have
+			// been made available even though the runner has finished
+			runner.Interrupt()
+		}
+	}
+
+	// Having notified that the context has been finished, we still need to
+	// wait for the rest of the results
+	for i := len(results); i < len(gr.runners); i++ {
+		result := <-ch
+		results[result.RunnerID] = result
+	}
```

Comment:
Wouldn't this last forever when something goes haywire? What if one service goes down without sending a response (eg panic)?

Reply:
Need to test what happens with a panic :+1:

Reply:
The service seems to crash:
```
root@dd7548f474cf:/# ocis collaboration server
panic: Oppsie, overslept

goroutine 73 [running]:
github.com/owncloud/ocis/v2/services/collaboration/pkg/command.Server.func2.5()
	/home/juan/src/ocis/ocis/services/collaboration/pkg/command/server.go:177 +0x38
github.com/owncloud/ocis/v2/ocis-pkg/runner.(*Runner).doTask(0xc0014f76a0, 0xc001546000, 0x0)
	/home/juan/src/ocis/ocis/ocis-pkg/runner/runner.go:91 +0x30
created by github.com/owncloud/ocis/v2/ocis-pkg/runner.(*Runner).RunAsync in goroutine 1
	/home/juan/src/ocis/ocis/ocis-pkg/runner/runner.go:77 +0xb0
root@dd7548f474cf:/# 
```

Relevant code piece is:
```
	gr.Add(runner.New("http", func() error {
		time.Sleep(15 * time.Second)
		panic("Oppsie, overslept")
		err := httpServer.Server().Start()
		if err == nil {
			err = <-httpCh // execution should block here
		}
		logger.Error().Err(err).Str("server", "http").Msg("Stopped")
		return err
	}, func() {
		logger.Error().Str("server", "http").Msg("shutting down server")
		httpCh <- httpServer.Server().Stop()
		close(httpCh)
	}))
	_ = gr.Run(ctx)
```
I assume the behavior will be the same if the panic happens inside the server, while handling a request.

I'm not sure if we need to do anything. Even if the runner catches the panic, the request causing the panic might be compromised already. The only positive thing is that we could request a graceful shutdown of the rest of the services.
It might depend on the specific service, so maybe we need to go on a case-by-case basis